Flaky-test: CompactionConcurrencyTest. testDisableCompactionConcurrently

### Search before reporting

- [x] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.


### Example failure

https://github.com/apache/pulsar/actions/runs/25842031561/job/75987701766?pr=25773

### Exception stacktrace

```text
Gradle suite > Gradle test > org.apache.pulsar.broker.service.persistent.CompactionConcurrencyTest > testDisableCompactionConcurrently FAILED
      org.awaitility.core.ConditionTimeoutException: Assertion condition expected [true] but found [false] within 10 seconds.
          at app//org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:167)
          at app//org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
          at app//org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
          at app//org.awaitility.core.ConditionFactory.until(ConditionFactory.java:985)
          at app//org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:769)
          at app//org.apache.pulsar.broker.service.persistent.CompactionConcurrencyTest.testDisableCompactionConcurrently(CompactionConcurrencyTest.java:73)
  
          Caused by:
          java.lang.AssertionError: expected [true] but found [false]
              at org.testng.Assert.fail(Assert.java:111)
              at org.testng.Assert.failNotEquals(Assert.java:1590)
              at org.testng.Assert.assertTrue(Assert.java:57)
              at org.testng.Assert.assertTrue(Assert.java:67)
              at org.apache.pulsar.broker.service.persistent.CompactionConcurrencyTest.lambda$testDisableCompactionConcurrently$3(CompactionConcurrencyTest.java:78)
  
  999 tests completed, 1 failed, 9 skipped
```

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/CompactionConcurrencyTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/CompactionConcurrencyTest.java
@@ -19,13 +19,16 @@
 package org.apache.pulsar.broker.service.persistent;
 
 import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
 import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.atomic.AtomicReference;
 import org.apache.bookkeeper.mledger.Position;
 import org.apache.pulsar.broker.service.SharedPulsarBaseTest;
+import org.apache.pulsar.client.admin.PulsarAdminException;
 import org.apache.pulsar.client.api.MessageId;
 import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.common.util.FutureUtil;
 import org.apache.pulsar.compaction.Compactor;
 import org.awaitility.Awaitility;
 import org.testng.annotations.Test;
@@ -56,27 +59,41 @@ public class CompactionConcurrencyTest extends SharedPulsarBaseTest {
         producer.newMessage().key("k0").value("v0").send();
         triggerCompactionAndWait(topicName);
         admin.topics().deleteSubscription(topicName, "s1");
+        PersistentTopic persistentTopic =
+                (PersistentTopic) getTopic(topicName, false).get().get();
 
         // Disable compaction.
         admin.topicPolicies().removeCompactionThreshold(topicName);
-        AtomicReference<CompletableFuture<Void>> f1 = new AtomicReference<CompletableFuture<Void>>();
-        AtomicReference<CompletableFuture<Void>> f2 = new AtomicReference<CompletableFuture<Void>>();
-        new Thread(() -> {
-            f1.set(admin.topics().deleteSubscriptionAsync(topicName, "__compaction"));
-        }).start();
-        new Thread(() -> {
-            f2.set(admin.topics().deleteSubscriptionAsync(topicName, "__compaction"));
-        }).start();
+        CompletableFuture<Long> originalCompaction = persistentTopic.currentCompaction;
+        CompletableFuture<Long> blockedCompaction = new CompletableFuture<>();
+        persistentTopic.currentCompaction = blockedCompaction;
+        try {
+            CompletableFuture<Void> firstDelete =
+                    admin.topics().deleteSubscriptionAsync(topicName, Compactor.COMPACTION_SUBSCRIPTION);
+            Awaitility.await().untilAsserted(() -> assertTrue(persistentTopic.disablingCompaction.get()));
 
-        // Verify: at least one of the requests should fail (the other may succeed or also fail
-        // with "not found" if the in-memory metadata store processes them sequentially).
-        Awaitility.await().untilAsserted(() -> {
-            assertTrue(f1.get() != null);
-            assertTrue(f2.get() != null);
-            assertTrue(f1.get().isDone());
-            assertTrue(f2.get().isDone());
-            assertTrue(f1.get().isCompletedExceptionally() || f2.get().isCompletedExceptionally());
-        });
+            CompletableFuture<Void> secondDelete =
+                    admin.topics().deleteSubscriptionAsync(topicName, Compactor.COMPACTION_SUBSCRIPTION);
+            Awaitility.await().untilAsserted(() -> assertTrue(secondDelete.isCompletedExceptionally()));
+            try {
+                secondDelete.join();
+                fail("The second concurrent compaction subscription delete should fail");
+            } catch (Exception ex) {
+                Throwable actEx = FutureUtil.unwrapCompletionException(ex);
+                assertTrue(actEx instanceof PulsarAdminException.PreconditionFailedException);
+            }
+
+            blockedCompaction.complete(0L);
+            Awaitility.await().untilAsserted(() -> {
+                assertTrue(firstDelete.isDone());
+                assertFalse(firstDelete.isCompletedExceptionally());
+                assertFalse(persistentTopic.disablingCompaction.get());
+            });
+            firstDelete.join();
+        } finally {
+            blockedCompaction.complete(0L);
+            persistentTopic.currentCompaction = originalCompaction;
+        }
 
         // cleanup.
         producer.close();

